crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3349,26 +3349,56 @@ impl<'db> TypeInferenceBuilder<'db> {
 
             // Walk up parent scopes looking for a possible enclosing scope that may have a
             // definition of this name visible to us (would be `LOAD_DEREF` at runtime.)
-            for (enclosing_scope_file_id, _) in self.index.ancestor_scopes(file_scope_id) {
+            let mut look_up_eagerly = scope.is_eager(db);
+            for (enclosing_scope_file_id, _) in self.index.ancestor_scopes(file_scope_id).skip(1) {
+                // Lazy scopes are "sticky": once we see a lazy scope we stop doing lookups
+                // eagerly, even if we encounter another eager enclosing scope later on.
+                let enclosing_scope_id = enclosing_scope_file_id.to_scope_id(db, current_file);
+                let look_up_eagerly_inner = look_up_eagerly;
+                look_up_eagerly = look_up_eagerly && enclosing_scope_id.is_eager(db);
```

Comment:
I update the flag for the next iteration here, up front, so that I don't have to remember to copy this line at each `continue` point

Reply:
No longer relevant